sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -1039,7 +1039,7 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
     val pythonEvals = collect(joinNode.get) {
       case p: BatchEvalPythonExec => p
     }
-    assert(pythonEvals.size == 2)
+    assert(pythonEvals.size == 4)
```

Comment:
@HyukjinKwon I'm not sure if this change can optimize python udf?

Reply:
Yeah, I don't think it's more efficient to have `BatchEvalPythonExec` more. It will require more Python executions which aren't trivial.

Reply:
I quickly checked:

```
== Physical Plan ==
*(3) Project [a#225, b#226, c#236, d#237]
+- *(3) BroadcastHashJoin [cast(pythonUDF0#256 as int)], [cast(pythonUDF0#257 as int)], Inner, BuildRight
   :- BatchEvalPython [udf(cast(a#225 as string))], [pythonUDF0#256]
   :  +- *(1) Project [_1#220 AS a#225, _2#221 AS b#226]
   :     +- *(1) Project [_1#220, _2#221]
   :        +- *(1) Filter isnotnull(cast(pythonUDF0#254 as int))
   :           +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#254]
   :              +- LocalTableScan [_1#220, _2#221]
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(cast(input[2, string, true] as int) as bigint))), [id=#140]
      +- BatchEvalPython [udf(cast(c#236 as string))], [pythonUDF0#257]
         +- *(2) Project [_1#231 AS c#236, _2#232 AS d#237]
            +- *(2) Project [_1#231, _2#232]
               +- *(2) Filter isnotnull(cast(pythonUDF0#255 as int))
                  +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#255]
                     +- LocalTableScan [_1#231, _2#232]
```

We should probably avoid inferring the is-not-null filter in this case.